### PR TITLE
Adds tensorflow support and async TF ros controller. 

### PR DESCRIPTION
**Pull request by bstadie**
This pull request adds tensorflow support. In particular, it allows badmm examples to utilize a tensorflow backend. The following additions are made:
- PolicyOptTf class for optimizing policy in tensorflow.
- TfPolicy class for running tensorlow policies forward.
- Auxiliary classes TfSolver and TfMap which allow reproducible specification of tensorflow models.
- ROS agent receives a new tensorflow controller, which allows for async control of the robot, with observations being published in the ROS CPP agent, received by the python agent, and then run forward though tensorflow in python. The resulting actions are published back to ROS where they are received by the CPP agent and used to directly set joint torques.
- In addition to ROS, TF policy optimization  is supported by Box2D and MuJoCo agents (I have tested them on my machine). 
- Add a tests folder and relevant tests for the newly added code.

### CODE REVIEW

File: experiments/pr2_tensorflow_example/hyperparams.py [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+    JOINT_VELOCITIES: 7,
+    END_EFFECTOR_POINTS: 3 * EE_POINTS.shape[0],
+    END_EFFECTOR_POINT_VELOCITIES: 3 * EE_POINTS.shape[0],
+    ACTION: 7,
+}
+
+PR2_GAINS = np.array([3.09, 1.08, 0.393, 0.674, 0.111, 0.152, 0.098])
+
+BASE_DIR = '/'.join(str.split(gps_filepath, '/')[:-2])
+EXP_DIR = BASE_DIR + '/../experiments/pr2_badmm_example/'
```
Comment (cbfinn) [inline experiments/pr2_tensorflow_example/hyperparams.py]: should be changed appropriately.

File: python/gps/agent/ros/agent_ros.py [hunk 2]
```diff
@@ -7,10 +9,12 @@
 from gps.agent.agent_utils import generate_noise, setup
 from gps.agent.config import AGENT_ROS
 from gps.agent.ros.ros_utils import ServiceEmulator, msg_to_sample, \
-        policy_to_msg
+        policy_to_msg, tf_policy_to_action_msg, tf_obs_msg_to_numpy
 from gps.proto.gps_pb2 import TRIAL_ARM, AUXILIARY_ARM
 from gps_agent_pkg.msg import TrialCommand, SampleResult, PositionCommand, \
-        RelaxCommand, DataRequest
+        RelaxCommand, DataRequest, TfActionCommand, TfObsData
+from gps.algorithm.policy.tf_policy import TfPolicy
```
Comment (cbfinn) [inline python/gps/agent/ros/agent_ros.py]: This import probably requires tensorflow to be installed, which is bad if someone wants to use AgentROS w/o tensorflow installed. Can you make sure that the examples all still run without tensorflow installed? (You can use a try/catch if needed)

File: python/gps/agent/ros/agent_ros.py [hunk 3]
```diff
@@ -155,11 +164,69 @@ def sample(self, policy, condition, verbose=True, save=True):
...
+                    should_stop = True
+        rospy.sleep(0.25)  # wait for finished trial to come in.
+        result = self._trial_service._subscriber_msg
+        return result  # the trial has completed. Here is its message.
+
+    def _get_new_action(self, policy, obs):
+        return policy.act(None, obs, None, None)
+
+    def _tf_callback(self, message):
+            self._tf_subscriber_msg = message
```
Comment (cbfinn) [inline python/gps/agent/ros/agent_ros.py]: nitpick - indentation is too big.

File: python/gps/algorithm/policy_opt/config_tf.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
+from gps.algorithm.policy_opt.tf_model_example import example_tf_network
+import os
+checkpoint_path = os.path.abspath(os.path.join(os.path.dirname(__file__),
+                                               '..', 'policy_opt/tf_checkpoint/policy_checkpoint.ckpt'))
+
+POLICY_OPT_TF = {
```
Comment (cbfinn) [inline python/gps/algorithm/policy_opt/config_tf.py]: Can you put this in policy_opt/config.py? Generally, we have been only using one config file per directory.

File: python/gps/algorithm/policy/tf_policy.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+        sess: tf session.
+        device_string: tf device string for running on either gpu or cpu.
+    """
+    def __init__(self, dU, obs_tensor, act_op, var, sess, device_string):
+        Policy.__init__(self)
+        self.dU = dU
+        self.obs_tensor = obs_tensor
+        self.act_op = act_op
+        self.sess = sess
+        self.device_string = device_string  # is it even worth running this on the gpu? Won't comm time dominate?
```
Comment (cbfinn) [inline python/gps/algorithm/policy/tf_policy.py]: For convoluational network policies, even with a batch size of 1, we've found big speed improvements from a GPU.
Remove comment?

**Comment by cbfinn** [comment]
This looks great Bradley! I'm about halfway through looking at it. I'll resume tonight.

File: python/gps/agent/ros/ros_utils.py [hunk 6]
```diff
@@ -4,9 +4,10 @@
 import rospy
 
 from gps.algorithm.policy.lin_gauss_policy import LinearGaussianPolicy
-from gps_agent_pkg.msg import ControllerParams, LinGaussParams, CaffeParams
+from gps_agent_pkg.msg import ControllerParams, LinGaussParams, TfParams, CaffeParams, TfActionCommand
 from gps.sample.sample import Sample
-from gps.proto.gps_pb2 import LIN_GAUSS_CONTROLLER, CAFFE_CONTROLLER
+from gps.proto.gps_pb2 import LIN_GAUSS_CONTROLLER, CAFFE_CONTROLLER, TF_CONTROLLER
+from gps.algorithm.policy.tf_policy import TfPolicy
```
Comment (cbfinn) [inline python/gps/agent/ros/ros_utils.py]: here's another place where tf is going to throw an error if it isn't installed

File: src/gps_agent_pkg/src/robotplugin.cpp [hunk 7]
```diff
@@ -532,3 +548,38 @@ void RobotPlugin::get_fk_solver(boost::shared_ptr<KDL::ChainFkSolverPos> &fk_sol
         ROS_ERROR("Unknown ArmType %i requested for joint encoder readings!",arm);
     }
 }
+
+
+
```
Comment (cbfinn) [inline src/gps_agent_pkg/src/robotplugin.cpp]: nitpick - can you remove all this space? should be 1 line of whitespace between functions.

File: tests/tests_tensorflow/test_policy_opt_tf.py [hunk 8]
```diff
@@ -0,0 +1,240 @@
+""" This file defines policy optimization for a tensorflow policy. """
```
Comment (cbfinn) [inline tests/tests_tensorflow/test_policy_opt_tf.py]: These tests look great.

**Comment by cbfinn** [comment]
Other than the two code comments above, this is good to go. Feel free to merge after fixing.

**Comment by bstadie** [comment]
I can not merge, as I do not have write access to this repo.

**Comment by cbfinn** [comment]
Just gave you write access for the future. I'll merge this PR now.

**Pull request merged**